// @loaders.gl/las/dist/esm/lib/parse-las.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { getMeshBoundingBox } from '@loaders.gl/loader-utils';
import { LASFile } from './laslaz-decoder';

function detectTwoByteColors(colorDepth, decoder, batchSize) {
  var twoByteColor;

  switch (colorDepth) {
    case 8:
      twoByteColor = false;
      break;

    case 16:
      twoByteColor = true;
      break;

    case 'auto':
      if (decoder.getPoint(0).color) {
        for (var i = 0; i < batchSize; i++) {
          var _decoder$getPoint = decoder.getPoint(i),
              color = _decoder$getPoint.color;

          if (color[0] > 255 || color[1] > 255 || color[2] > 255) {
            twoByteColor = true;
          }
        }
      }

      break;

    default:
      console.warn('las: illegal value for options.las.colorDepth');
      break;
  }

  return twoByteColor;
}

export default function parseLAS(arraybuffer) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var pointIndex = 0;
  var positions;
  var colors;
  var intensities;
  var classifications;
  var originalHeader;
  var result = {};
  var onProgress = options.onProgress;

  var _ref = options.las || {},
      skip = _ref.skip,
      colorDepth = _ref.colorDepth,
      fp64 = _ref.fp64;

  parseLASChunked(arraybuffer, skip, function (decoder, header) {
    if (!originalHeader) {
      originalHeader = header;
      var total = header.totalToRead;
      var PositionsType = fp64 ? Float64Array : Float32Array;
      positions = new PositionsType(total * 3);
      colors = header.pointsFormatId >= 2 ? new Uint8Array(total * 4) : null;
      intensities = new Uint16Array(total);
      classifications = new Uint8Array(total);
      Object.assign(result, {
        loaderData: {
          header: header
        },
        mode: 0,
        attributes: {
          POSITION: {
            value: positions,
            size: 3
          },
          intensity: {
            value: intensities,
            size: 1
          },
          classification: {
            value: classifications,
            size: 1
          }
        }
      });

      if (colors) {
        result.attributes.COLOR_0 = {
          value: colors,
          size: 4
        };
      }
    }

    var batchSize = decoder.pointsCount;

    var _header$scale = _slicedToArray(header.scale, 3),
        scaleX = _header$scale[0],
        scaleY = _header$scale[1],
        scaleZ = _header$scale[2],
        _header$offset = _slicedToArray(header.offset, 3),
        offsetX = _header$offset[0],
        offsetY = _header$offset[1],
        offsetZ = _header$offset[2];

    var twoByteColor = detectTwoByteColors(colorDepth, decoder, batchSize);

    for (var i = 0; i < batchSize; i++) {
      var _decoder$getPoint2 = decoder.getPoint(i),
          position = _decoder$getPoint2.position,
          color = _decoder$getPoint2.color,
          intensity = _decoder$getPoint2.intensity,
          classification = _decoder$getPoint2.classification;

      positions[pointIndex * 3] = position[0] * scaleX + offsetX;
      positions[pointIndex * 3 + 1] = position[1] * scaleY + offsetY;
      positions[pointIndex * 3 + 2] = position[2] * scaleZ + offsetZ;

      if (color) {
        if (twoByteColor) {
          colors[pointIndex * 4] = color[0] / 256;
          colors[pointIndex * 4 + 1] = color[1] / 256;
          colors[pointIndex * 4 + 2] = color[2] / 256;
        } else {
          colors[pointIndex * 4] = color[0];
          colors[pointIndex * 4 + 1] = color[1];
          colors[pointIndex * 4 + 2] = color[2];
        }

        colors[pointIndex * 4 + 3] = 255;
      }

      intensities[pointIndex] = intensity;
      classifications[pointIndex] = classification;
      pointIndex++;
    }

    if (onProgress) {
      onProgress(Object.assign({
        header: {
          vertexCount: header.totalRead
        },
        progress: header.totalRead / header.totalToRead
      }, result));
    }
  });
  result.header = {
    vertexCount: originalHeader.totalToRead,
    boundingBox: getMeshBoundingBox(result.attributes)
  };
  return result;
}
export function parseLASChunked(rawData, skip, onParseData) {
  var dataHandler = new LASFile(rawData);

  try {
    dataHandler.open();
    var header = dataHandler.getHeader();
    var Unpacker = dataHandler.getUnpacker();
    var totalToRead = Math.ceil(header.pointsCount / Math.max(1, skip));
    header.totalToRead = totalToRead;
    var totalRead = 0;

    while (true) {
      var chunk = dataHandler.readData(1000 * 100, 0, skip);
      totalRead += chunk.count;
      header.totalRead = totalRead;
      header.versionAsString = chunk.versionAsString;
      header.isCompressed = chunk.isCompressed;
      var unpacker = new Unpacker(chunk.buffer, chunk.count, header);
      onParseData(unpacker, header);

      if (!chunk.hasMoreData || totalRead >= totalToRead) {
        break;
      }
    }
  } catch (e) {
    throw e;
  } finally {
    dataHandler.close();
  }
}
